File: domain/src/main/java/com/study/platform/domain/notification/model/FailedNotification.java
package com.study.platform.domain.notification.model;

import com.study.platform.global.entity.BaseTimeEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.UUID;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "failed_notifications")
public class FailedNotification extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "BINARY(16)")
    private UUID id;

    @Column(nullable = false, columnDefinition = "BINARY(16)")
    private UUID receiverId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private NotificationType type;

    @Column(nullable = false)
    private String message;

    @Column(columnDefinition = "BINARY(16)")
    private UUID targetId;

    @Column(columnDefinition = "text")
    private String failureReason;

    public static FailedNotification from(NotificationEvent event, String failureReason) {
        FailedNotification entity = new FailedNotification();
        entity.receiverId = event.receiverId();
        entity.type = event.type();
        entity.message = event.message();
        entity.targetId = event.targetId();
        entity.failureReason = failureReason;
        return entity;
    }
}

File: domain/src/main/kotlin/com/study/platform/domain/notification/model/FailedNotification.kt
package com.study.platform.domain.notification.model

import com.study.platform.global.entity.BaseTimeEntity
import jakarta.persistence.*
import java.util.UUID

@Entity
@Table(name = "failed_notifications")
class FailedNotification : BaseTimeEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "BINARY(16)")
    var id: UUID? = null

    @Column(nullable = false, columnDefinition = "BINARY(16)")
    var receiverId: UUID? = null

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    var type: NotificationType? = null

    @Column(nullable = false)
    var message: String = ""

    @Column(columnDefinition = "BINARY(16)")
    var targetId: UUID? = null

    @Column(columnDefinition = "text")
    var failureReason: String? = null

    companion object {
        fun from(event: NotificationEvent, failureReason: String): FailedNotification =
            FailedNotification().also {
                it.receiverId = event.receiverId
                it.type = event.type
                it.message = event.message
                it.targetId = event.targetId
                it.failureReason = failureReason
            }
    }
}
